src/error.rs

```diff
@@ -44,6 +56,40 @@ impl<T> OptionalExtension<T> for Result<T, ConnectionError> {
     }
 }
 
+impl<T> OptionalExtension<T> for Result<T, RunError<DieselError>> {
+    fn optional(self) -> Result<Option<T>, ConnectionError> {
+        let self_as_query_result: diesel::QueryResult<T> = match self {
+            Ok(value) => Ok(value),
+            Err(RunError::User(error_kind)) => Err(error_kind),
+            Err(RunError::RuntimeShutdown) => return Err(ConnectionError::RuntimeShutdown),
+        };
+
+        self_as_query_result
+            .optional()
+            .map_err(ConnectionError::Query)
+    }
+}
+
+/// Errors encountered while running a function on a connection pool.
+#[derive(Error, Debug, Clone, Copy, Eq, PartialEq)]
+pub enum RunError<E> {
+    #[error("error in user code")]
+    User(#[from] E),
```

Comment:
IMO, this should probably be `#[error(transparent)]` --- I don't think there's much value in replacing the formatted representation of the underlying error with the very opaque "error in user code", and only displaying the underlying error if the user displays the cause chain.
```suggestion
    #[error(transparent)]
    User(#[from] E),
```

Reply:
Done, I swore I had changed this before putting it up...